src/cls/base.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EventEmitter} from 'events';
+
+// tslint:disable-next-line:no-any
+export type Func<T = void> = (...args: any[]) => T;
+
+/**
+ * An interface that represents a background mechanism which is capable of
+ * storing, propagating and retrieving arbitrary continuation-local data (also
+ * called "context").
+ *
+ * A continuation refers to a logical tree of execution paths; a function passed
+ * to CLS#runWithCurrentContext is considered the root of such a tree, and
+ * subsequent child nodes are functions that are "triggered" because of
+ * asynchronous operations started by their parent (for example, a function that
+ * calls fs.readFile "triggers" the callback that is passed to it).
+ * The exact definition of "trigger" is implementation-dependent.
+ *
+ * CLS stands for continuation-local storage.
+ */
+export interface CLS<Context extends {}> {
+  /**
+   * Returns whether this continuation-local storage mechanism is enabled.
+   */
+  isEnabled(): boolean;
+
+  /**
+   * Enables this instance.
+   */
+  enable(): void;
```

Comment:
As stated elsewhere, I would prefer if instantiation subsumed `enable`, and `disable` was replaced with `destroy`.  It would be better to create instances when needed rather than having long lived instanced that could be enabled / disabled. Less global state is better.

Reply:
`enable`/`disable` don't necessarily have to do with long-livedness of objects -- even if we are creating short-lived instances I believe there should be separation between the instantiation of an object, and the mutation of global state. The latter doesn't belong in the constructor, and I think enabling a set of async hooks falls into the latter category. So I would prefer that `enable` is separate from the constructor.